subsys/net/ip/utils.c

```diff
@@ -518,3 +518,230 @@ bool net_header_fits(struct net_pkt *pkt, u8_t *hdr, size_t hdr_size)
+
+		if ((ptr + 1) < (str + str_len) && *(ptr + 1) == ':') {
+			len = str_len - end;
+
+			/* Re-use the ipaddr buf for port conversion */
+			memcpy(ipaddr, ptr + 2, len);
+			ipaddr[len] = '\0';
+
+			ret = convert_port(ipaddr, &port);
+			if (!ret) {
+				return false;
+			}
+
+			net_sin6(addr)->sin6_port = htons(port);
+
+			NET_DBG("IPv6 host %s port %d",
+				net_addr_ntop(AF_INET6,
+					      &net_sin6(addr)->sin6_addr,
+					      ipaddr, sizeof(ipaddr) - 1),
+				port);
+		} else {
+			NET_DBG("IPv6 host %s",
+				net_addr_ntop(AF_INET6,
+					      &net_sin6(addr)->sin6_addr,
+					      ipaddr, sizeof(ipaddr) - 1));
+		}
+
+		return true;
+#else
+		return false;
+#endif /* CONFIG_NET_IPV6 */
+	}
+
+	for (count = i = 0; str[i] && i < str_len; i++) {
+		if (str[i] == ':') {
+			count++;
+		}
+	}
+
+	if (count == 1) {
+#if defined(CONFIG_NET_IPV4)
+		/* IPv4 address with port number */
+		char ipaddr[NET_IPV4_ADDR_LEN + 1];
+		struct in_addr *addr4;
+		int end, len;
+
+		ptr = memchr(str, ':', str_len);
+		if (!ptr) {
+			return false;
+		}
+
+		len = min(NET_IPV4_ADDR_LEN, str_len);
+		end = min(len, ptr - str);
+
+		memcpy(ipaddr, str, end);
+		ipaddr[end] = '\0';
+
+		addr4 = &net_sin(addr)->sin_addr;
+
+		net_sin6(addr)->sin6_family = AF_INET;
+
+		ret = net_addr_pton(AF_INET, ipaddr, addr4);
+		if (ret < 0) {
+			return false;
+		}
+
+		memcpy(ipaddr, ptr + 1, str_len - end);
+		ipaddr[str_len - end] = '\0';
+
+		ret = convert_port(ipaddr, &port);
+		if (!ret) {
+			return false;
+		}
+
+		net_sin(addr)->sin_port = htons(port);
+
+		NET_DBG("IPv4 host %s port %d",
+			net_addr_ntop(AF_INET,
+				      &net_sin(addr)->sin_addr,
+				      ipaddr, sizeof(ipaddr) - 1),
+			port);
+		return true;
+#else
+		return false;
+#endif /* CONFIG_NET_IPV4 */
+	}
+
+#if defined(CONFIG_NET_IPV4) && defined(CONFIG_NET_IPV6)
+	char ipaddr[INET6_ADDRSTRLEN + 1];
+	int end, len;
+
+	len = min(NET_IPV4_ADDR_LEN, str_len);
+	end = min(len, strlen(str));
+
+	memcpy(ipaddr, str, end);
+	ipaddr[end] = '\0';
+
+	/* First try IPv4 address */
+	ret = net_addr_pton(AF_INET, ipaddr, &net_sin(addr)->sin_addr);
+	if (ret < 0 && ret != -EINVAL) {
+		return false;
+	}
+
+	if (!ret) {
+		/* Caller must set the port if it could not be figured
+		 * out from address string.
+		 */
+		net_sin(addr)->sin_family = AF_INET;
+
+		NET_DBG("IPv4 host %s", str);
+
+	} else if (ret == -EINVAL) {
+		/* Then the address must be IPv6 based */
+		len = min(NET_IPV6_ADDR_LEN, str_len);
+		end = min(len, strlen(str));
+
+		memcpy(ipaddr, str, end);
+		ipaddr[end] = '\0';
+
+		ret = net_addr_pton(AF_INET6, ipaddr,
+				    &net_sin6(addr)->sin6_addr);
+		if (ret < 0) {
+			return false;
+		}
+
+		net_sin6(addr)->sin6_family = AF_INET6;
+
+		NET_DBG("IPv6 host %s", str);
+	} else {
+		return false;
+	}
+#endif
+
+#if defined(CONFIG_NET_IPV4) && !defined(CONFIG_NET_IPV6)
+	char ipaddr[NET_IPV4_ADDR_LEN + 1];
+	int end, len;
+
+	len = min(NET_IPV4_ADDR_LEN, str_len);
```

Comment:
Do we have `alloca()` support?  If so, these things copying `str` into `ipaddr` could become just a call to `strndupa()`.   [Soletta](https://github.com/solettaproject/soletta/blob/master/src/lib/common/sol-missing.h#L30-L38) has a simple implementation that you can borrow (same license, Intel owns the copyright).

Reply:
Lets not introduce yet another dependecy here :)